hudi-spark-datasource/hudi-spark-common/src/main/java/org/apache/hudi/HoodieSparkRecordMerger.java

```diff
@@ -38,44 +37,87 @@ public String getMergingStrategy() {
+                                                  Schema oldSchema,
+                                                  Option<HoodieRecord> newer,
+                                                  Schema newSchema,
+                                                  TypedProperties props) throws IOException {
+    boolean isValidNew = isValid(newer, newSchema, props);
+    boolean isValidOld = isValid(older, oldSchema, props);
 
-    if (newer instanceof HoodieSparkRecord) {
-      HoodieSparkRecord newSparkRecord = (HoodieSparkRecord) newer;
-      if (newSparkRecord.isDeleted()) {
-        // Delete record
-        return Option.empty();
-      }
+    if (!isValidOld && !isValidNew) {
+      return handleTrivialCase(older, oldSchema, newer, newSchema, props);
+    } else if (isValidOld && !isValidNew) {
+      return handleDeleteCase(older, oldSchema, newer, newSchema, props);
+    } else if (!isValidOld) {
+      return handleInsertCase(older, oldSchema, newer, newSchema, props);
     } else {
-      if (newer.getData() == null) {
-        // Delete record
-        return Option.empty();
-      }
-    }
-
-    if (older instanceof HoodieSparkRecord) {
-      HoodieSparkRecord oldSparkRecord = (HoodieSparkRecord) older;
-      if (oldSparkRecord.isDeleted()) {
-        // use natural order for delete record
-        return Option.of(Pair.of(newer, newSchema));
-      }
-    } else {
-      if (older.getData() == null) {
-        // use natural order for delete record
-        return Option.of(Pair.of(newer, newSchema));
-      }
-    }
-    if (older.getOrderingValue(oldSchema, props).compareTo(newer.getOrderingValue(newSchema, props)) > 0) {
-      return Option.of(Pair.of(older, oldSchema));
-    } else {
-      return Option.of(Pair.of(newer, newSchema));
+      return handleUpdateCase(older, oldSchema, newer, newSchema, props);
     }
   }
 
   @Override
   public HoodieRecordType getRecordType() {
     return HoodieRecordType.SPARK;
   }
+
+  private boolean isValid(Option<HoodieRecord> record, Schema schema, TypedProperties props) throws IOException {
+    return record.isPresent()
+        && !(record.get() instanceof HoodieEmptyRecord)
+        && !(record.get().isDelete(schema, props))
+        && record.get().getRecordType() == HoodieRecordType.SPARK;
+  }
+
+  /**
+   * Handle trivial case.
+   */
+  Option<Pair<HoodieRecord, Schema>> handleTrivialCase(
+      Option<HoodieRecord> olderIgnored,
+      Schema oldSchemaIgnored,
+      Option<HoodieRecord> newerIgnored,
+      Schema newSchemaIgnored,
+      TypedProperties propsIgnored) throws IOException {
+    return Option.empty();
+  }
+
+  /**
+   * Handle delete case.
+   */
+  Option<Pair<HoodieRecord, Schema>> handleDeleteCase(
+      Option<HoodieRecord> olderIgnored,
+      Schema oldSchemaIgnored,
+      Option<HoodieRecord> newerIgnored,
+      Schema newSchemaIgnored,
+      TypedProperties propsIgnored) throws IOException {
+    return Option.empty();
```

Comment:
For the delete case, should the `HoodieEmptyRecord` be returned?

Reply:
Either one should work, but checking `option.isPresent()` should be more efficient than `instanceOf`. Wdyt?

Reply:
It depends on the merging case.  When merging the records from log records with the ordering field, we have to use `HoodieEmptyRecord` to include the record key and ordering field to indicate that this is a delete record and used for merging with deletes/updates from other log files.